### PR TITLE
LIZA-129: Graph: Рефакторинг 

### PR DESCRIPTION
**Pull request by ol00lo**

### CODE REVIEW

File: graph/lib/data_node.hpp [hunk 1]
```diff
@@ -7,17 +7,19 @@ namespace g
 class DataNode : public INode
 {
 public:
-    DataNode(std::string nodename) : INode(nodename), _value({0}) {};
-    void serialize_spec(nlohmann::json& js) const override;
-    void deserialize_spec(const nlohmann::json&, std::string copy_word = "_copy") override;
-    Tensor get_value() override;
+    DataNode(std::string nodename) : INode("DataNode",  nodename), value_({0}) {};
+
+	void serialize_spec(nlohmann::json& js) const override;
```
Comment (kalininei) [inline graph/lib/data_node.hpp:12]: формат

File: graph/lib/i_functional_node.hpp [hunk 2]
```diff
@@ -9,41 +9,38 @@ namespace g
 class IFunctionalNode : public INode
 {
 public:
-    IFunctionalNode(std::string nodename = "") : INode(nodename), _value({0})
-    {
-    }
-    using callback_t = std::function<void(IFunctionalNode*)>;
+    IFunctionalNode(std::string classname, std::string nodename = "") : INode(classname, nodename), value_({0}) {};
 
+	using callback_t = std::function<void(IFunctionalNode*)>;
```
Comment (kalininei) [inline graph/lib/i_functional_node.hpp:14]: формат

File: graph/lib/model.cpp [hunk 3]
```diff
@@ -9,7 +9,6 @@
 using namespace g;
 namespace
 {
-using pnode_t = INode::ptr_t;
 pnode_t find_node_by_name(std::unordered_map<std::string, pnode_t> all_nodes, const std::string& name)
```
Comment (kalininei) [inline graph/lib/model.cpp:12]: &
И она вроде не используется?

File: graph/lib/model.cpp [hunk 4]
```diff
@@ -119,5 +124,16 @@ Model Model::load(const std::string& filename)
...
+    return res;
 }
+
+void Model::save_name(pnode_t node)
+{
+    std::string nodename = node->nodename();
+    if (names_.find(nodename) == names_.end())
+        names_.insert(nodename);
+    else
+        node->rename();
```
Comment (kalininei) [inline graph/lib/model.cpp:138]: не, не надо переименовывать автоматически. Надо бросать ошибку.
А если хотите сделать `rename`, то сделайте её с аргументом -- новым именем

File: graph/lib/model.cpp [hunk 5]
```diff
@@ -21,77 +20,83 @@ pnode_t find_node_by_name(std::unordered_map<std::string, pnode_t> all_nodes, co
...
+void Model::set_param_nodes(const std::vector<std::shared_ptr<DataNode>>& p)
 {
-    for (int i = 0; i < _param_nodes.size(); i++)
+    for (int i = 0; i < param_nodes_.size(); i++)
     {
-        _param_nodes[i]->set_value(p[i]->get_value());
+        param_nodes_[i]->set_value(p[i]->value());
     }
 }
 void Model::add_into_inter(pnode_t node)
```
Comment (kalininei) [inline graph/lib/model.cpp:81]: смените название, если уж `inter_nodes` убрали.

File: graph/lib/model.cpp [hunk 6]
```diff
@@ -21,77 +20,83 @@ pnode_t find_node_by_name(std::unordered_map<std::string, pnode_t> all_nodes, co
...
+        nodes_.push_back(node);
         auto dataNode = std::dynamic_pointer_cast<DataNode>(node);
         if (dataNode != nullptr)
         {
-            _param_nodes.push_back(dataNode);
+            param_nodes_.push_back(dataNode);
         }
     }
-    for (const auto& prev : node->get_prev())
+    for (const auto& prev : node->prev_nodes())
```
Comment (kalininei) [inline graph/lib/model.cpp:99]: А разве возможна ситуация, что узел `not_in_inter`, но при этом требует обхода его детей?
Не нужно этот цикл в верхний иф унести?

File: graph/lib/optimization_worker.cpp [hunk 7]
```diff
@@ -4,26 +4,29 @@ using namespace g;
...
+    optimizer_->set_param_nodes(graph_.param_nodes());
 }
 
 double OptimizationWorker::train(const std::vector<Tensor>& inputs, const std::vector<Tensor>& gt)
 {
-    double res = compute_loss(inputs, gt);
-    auto grads = _graph.get_gradients();
-    _optimizer->apply(grads);
-    return res;
+    double loss = compute_loss(inputs, gt);
```
Comment (kalininei) [inline graph/lib/optimization_worker.cpp:13]: Решили не сдвигать в конец? Почему?

File: graph/lib/optimization_worker.cpp [hunk 8]
```diff
@@ -4,26 +4,29 @@ using namespace g;
...
+    optimizer_->apply(grads);
+    log().info("Train loss: {}", loss);
+    return loss;
 }
 
 double OptimizationWorker::validate(const std::vector<Tensor>& inputs, const std::vector<Tensor>& gt)
 {
-    return compute_loss(inputs, gt);
+    double loss = compute_loss(inputs, gt);
+    log().info("Valid loss: {}", loss);
```
Comment (kalininei) [inline graph/lib/optimization_worker.cpp:23]: Val loss.
Если хотите форматирование сохранить, лучше пробелы после `:` поставьте

File: graph/lib/tensor.hpp [hunk 9]
```diff
@@ -66,7 +69,7 @@ struct adl_serializer<g::Tensor>
 {
     static void to_json(json& j, const g::Tensor& t)
     {
-        j = json{{"value", t.data()}, {"shape", t.get_shape()}};
+        j = json{{"value", t.data()}, {"shape", t.shape()}};
```
Comment (kalininei) [inline graph/lib/tensor.hpp:72]: здесь бы тоже шейп в одну строку

File: graph/tests/tests.cpp
Comment (kalininei) [inline graph/tests/tests.cpp:1]: Напишите тест про то, как вы создаёте модель из узлов с одинаковыми именами и у вас бросается ошибка.

File: graph/lib/i_functional_node.hpp (hunk 2 above)
Comment (kalininei) [reply graph/lib/i_functional_node.hpp:14]: и ещё в остальных файлах проверьте

File: graph/lib/data_node.hpp (hunk 1 above)
Comment (ol00lo) [reply graph/lib/data_node.hpp]: ✔️

File: graph/lib/i_functional_node.hpp (hunk 2 above)
Comment (ol00lo) [reply graph/lib/i_functional_node.hpp]: ✔️

File: graph/lib/model.cpp (hunk 3 above)
Comment (ol00lo) [reply graph/lib/model.cpp]: ✔️

File: graph/lib/model.cpp (hunk 4 above)
Comment (ol00lo) [reply graph/lib/model.cpp]: ✔️

File: graph/lib/model.cpp (hunk 5 above)
Comment (ol00lo) [reply graph/lib/model.cpp]: ✔️

File: graph/lib/model.cpp (hunk 6 above)
Comment (ol00lo) [reply graph/lib/model.cpp]: ✔️

File: graph/lib/optimization_worker.cpp (hunk 7 above)
Comment (ol00lo) [reply graph/lib/optimization_worker.cpp:13]: Там один дополнительный раз будет считаться value в каждом узле. И у меня не получилось придумать, как этого избежать

File: graph/lib/optimization_worker.cpp (hunk 8 above)
Comment (ol00lo) [reply graph/lib/optimization_worker.cpp]: ✔️

File: graph/tests/tests.cpp
Comment (ol00lo) [reply graph/tests/tests.cpp:1]: ✔️

File: graph/lib/optimization_worker.cpp (hunk 7 above)
Comment (kalininei) [reply graph/lib/optimization_worker.cpp:13]: Мне видится, что после опускания вычисления лосса вниз, последовательность будет такая:
эпоха 1:
  - считаем градиент (пересчитываем value)
  - меняем модель
  - считаем лосс (пересчитываем value)
 эпоха 2:
  - считаем градиент (value из кэша)
  - меняем модель
  - считаем лосс (пересчитываем value)
  эпоха 3:
  - считаем градиент (value из кэша)
  - меняем модель
  - считаем лосс (пересчитываем value)

таким образом на каждой эпохе кроме первой value пересчитывается один раз.
Покажите, где я не прав.

File: graph/lib/tensor.hpp (hunk 9 above)
Comment (kalininei) [reply graph/lib/tensor.hpp]: Давайте для shape тоже adl_serializer сделаем тогда уж.

File: graph/lib/data_node.hpp (hunk 1 above)
Comment (kalininei) [reply graph/lib/data_node.hpp]: Осталось как минимум в model_test.cpp
У вас же скрипт есть, который во всех файлах должен править форматирование?

File: graph/lib/optimization_worker.cpp (hunk 7 above)
Comment (ol00lo) [reply graph/lib/optimization_worker.cpp]: ✔️

File: graph/lib/tensor.hpp (hunk 9 above)
Comment (ol00lo) [reply graph/lib/tensor.hpp]: ✔️

Comment (kalininei) [reply graph/lib/tensor.hpp]: Не, магия адл в том, чтобы не использовать конверсию, а просто иметь возможность писать
```cpp
//to_json
j = json{{"value", t.data()}, {"shape", t.shape()}};   // not json(t.shape())
//from json
g::Shape sh = j.at("shape").get<g::Shape>();          // not get<std::string>();
```
Кстати, конеструктор создания шейпа от строки уберите. Он в этой парадигме не нужен и только путает.
Да, там вам понадобится пару вещей исправить. В частности добаить дефолтный конструктор к шейпу и еще ошибка там где-то была.

File: graph/lib/data_node.hpp (hunk 1 above)
Comment (kalininei) [reply graph/lib/data_node.hpp]: Так и не исправили файл model_test.cpp))

File: graph/lib/tensor.hpp (hunk 9 above)
Comment (kalininei) [reply graph/lib/tensor.hpp]: И в тесте [tensor_serialize] тоже `get<std::string>` замените

File: graph/lib/data_node.hpp (hunk 1 above)
Comment (ol00lo) [reply graph/lib/data_node.hpp]: ✔️

File: graph/lib/tensor.hpp (hunk 9 above)
Comment (ol00lo) [reply graph/lib/tensor.hpp:72]: ✔️
